File: Grind75/68-SerializeDeserializeBinaryTree/cpp/SerializeDeserializeBinaryTree.cpp
#include "SerializeDeserializeBinaryTree.h"
#include <queue>
#include <string>
#include <Tree.h>
#include <vector>

using namespace std;
using namespace BST;

string Codec::serialize(TreeNode* root) {
  if (!root)
    return NULLSTR;
  queue<TreeNode*> nodeQ;
  nodeQ.push(root);

  string serialized = "";
  while (!nodeQ.empty()) {
    TreeNode *node = nodeQ.front();
    nodeQ.pop();
    if (!node) {
      serialized += NULLSTR + SEP;
      continue;
    }
    serialized += to_string(node->val) + SEP;
    nodeQ.push(node->left);
    nodeQ.push(node->right);
  }

  while (!isdigit(serialized.back()))
    serialized.pop_back();

  return serialized;
}

TreeNode* Codec::deserialize(string data) {
  data.push_back(',');
  vector<TreeNode*> nodes;
  string curNode = "";

  for (const char &c : data) {
    if (c == ',') {
      TreeNode *node = nullptr;
      if (curNode != NULLSTR)
        node = new TreeNode(atoi(curNode.c_str()));
      nodes.push_back(node);
      curNode.clear();
    } else {
      curNode.push_back(c);
    }
  }
  if (nodes.empty())
    return nullptr;

  vector<uint> holes(nodes.size(), 0);
  for (uint i = 0; i < nodes.size(); i++) {
    if (i > 0)
      holes[i] = holes[i-1];
    if (!nodes[i]) {
      holes[i]++;
      continue;
    }
    uint left  = 2 * (i - holes[i]) + 1;
    if (left < nodes.size())
      nodes[i]->left = nodes[left];
    uint right = 2 * (i - holes[i]) + 2;
    if (right < nodes.size())
      nodes[i]->right = nodes[right];
  }
  return nodes[0];
}

File: Grind75/68-SerializeDeserializeBinaryTree/python/SerializeDeserializeBinaryTree.py
#!/bin/python3

from __future__ import annotations
import sys
sys.path.append('../../Utility/python')
from PrettyPrint import *
from Colors import *
from Tree import *

class Codec:
  def __init__(self) -> None:
    self.NULLSTR: str = 'null'
    self.SEP: str = ','

  def serialize(self, root: Optional[TreeNode]) -> str:
    def preorder(node: Optional[TreeNode], prefix: List[str]) -> None:
      if node is None:
        prefix.append(self.NULLSTR)
        return
      prefix.append(str(node.val))
      preorder(node.left, prefix)
      preorder(node.right, prefix)
    nodes: List[str] = []
    preorder(root, nodes)
    return self.SEP.join(nodes)

  def deserialize(self, data: str) -> Optional[TreeNode]:
    nodes = iter(data.split(self.SEP))
    def linkNodes() -> Optional[TreeNode]:
      nodeStr = next(nodes)
      if nodeStr == self.NULLSTR:
        return None
      node = TreeNode(int(nodeStr))
      node.left = linkNodes()
      node.right = linkNodes()
      return node
    return linkNodes()


if __name__ == '__main__':
  codec = Codec()
  t = makeTree([
    TreeNode(1),
    TreeNode(2),TreeNode(3),
    None, None, TreeNode(4),TreeNode(5),
    None, None, None, None, TreeNode(6),TreeNode(7),
  ])

  s = codec.serialize(t)
  print('Serialized: ', quote(s))
  d = codec.deserialize(s)
  print('Deserialized:')
  printTree(d)

